provider/pkg/provider/environment_version_tags.go

```diff
@@ -0,0 +1,162 @@
+	serde.FromPropertyMap(inputMap, structTagKey, &input)
+	return input
+}
+
+func (evt *PulumiServiceEnvironmentVersionTagResource) Name() string {
+	return "pulumiservice:index:EnvironmentVersionTag"
+}
+
+func (evt *PulumiServiceEnvironmentVersionTagResource) Diff(req *pulumirpc.DiffRequest) (*pulumirpc.DiffResponse, error) {
+	olds, err := plugin.UnmarshalProperties(req.GetOlds(), plugin.MarshalOptions{KeepUnknowns: false, SkipNulls: true})
+	if err != nil {
+		return nil, err
+	}
+
+	news, err := plugin.UnmarshalProperties(req.GetNews(), plugin.MarshalOptions{KeepUnknowns: true, SkipNulls: true})
+	if err != nil {
+		return nil, err
+	}
+
+	diffs := olds.Diff(news)
+	if diffs == nil {
+		return &pulumirpc.DiffResponse{
+			Changes: pulumirpc.DiffResponse_DIFF_NONE,
+		}, nil
+	}
+
+	dd := plugin.NewDetailedDiffFromObjectDiff(diffs, false)
+
+	detailedDiffs := map[string]*pulumirpc.PropertyDiff{}
+	for k, v := range dd {
+		v.Kind = v.Kind.AsReplace()
+		detailedDiffs[k] = &pulumirpc.PropertyDiff{
+			Kind:      pulumirpc.PropertyDiff_Kind(v.Kind),
+			InputDiff: v.InputDiff,
+		}
+	}
+
+	return &pulumirpc.DiffResponse{
+		Changes:             pulumirpc.DiffResponse_DIFF_SOME,
+		DetailedDiff:        detailedDiffs,
+		DeleteBeforeReplace: true,
+		HasDetailedDiff:     true,
+	}, nil
+}
+
+func (evt *PulumiServiceEnvironmentVersionTagResource) Delete(req *pulumirpc.DeleteRequest) (*pbempty.Empty, error) {
+	ctx := context.Background()
+	var input PulumiServiceEnvironmentVersionTagInput
+	err := serde.FromProperties(req.GetProperties(), structTagKey, &input)
+	if err != nil {
+		return nil, err
+	}
+	err = evt.client.DeleteEnvironmentRevisionTag(ctx, input.Organization, input.Environment, input.TagName)
+	if err != nil {
+		return nil, err
+	}
+	return &pbempty.Empty{}, nil
+}
+
+func (evt *PulumiServiceEnvironmentVersionTagResource) Create(req *pulumirpc.CreateRequest) (*pulumirpc.CreateResponse, error) {
+	ctx := context.Background()
+	var input PulumiServiceEnvironmentVersionTagInput
+	err := serde.FromProperties(req.GetProperties(), structTagKey, &input)
+	if err != nil {
+		return nil, err
+	}
+	err = evt.client.CreateEnvironmentRevisionTag(ctx, input.Organization, input.Environment, input.TagName, &input.Revision)
+	if err != nil {
+		return nil, err
+	}
+	return &pulumirpc.CreateResponse{
+		Id:         path.Join(input.Organization, input.Environment, input.TagName, strconv.Itoa(input.Revision)),
+		Properties: req.GetProperties(),
+	}, nil
+}
+
+func (evt *PulumiServiceEnvironmentVersionTagResource) Check(req *pulumirpc.CheckRequest) (*pulumirpc.CheckResponse, error) {
+	return &pulumirpc.CheckResponse{Inputs: req.News, Failures: nil}, nil
+}
+
+func (evt *PulumiServiceEnvironmentVersionTagResource) Update(req *pulumirpc.UpdateRequest) (*pulumirpc.UpdateResponse, error) {
+	// all updates are destructive, so we just call Create.
+	return nil, fmt.Errorf("unexpected call to update, expected create to be called instead")
```

Comment:
We definitely don't want tags to always be replaced. Every change to a tag besides its name should be an update.

Reply:
I followed the way stack tags are implemented [here](https://github.com/pulumi/pulumi-pulumiservice/blob/f13da7e94e5f83093da265d3a3c7c4d842375ba9/provider/pkg/provider/stack_tags.go#L133)

I'll rework it to have updates